pkg/reconciler/pubsub/reconciler.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"context"
+	"errors"
+
+	"github.com/knative/pkg/kmeta"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/batch/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/apis/pubsub/v1alpha1"
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/pubsub/operations"
+	"github.com/GoogleCloudPlatform/cloud-run-events/pkg/reconciler"
+)
+
+const (
+	// ReconcilerName is the name of the reconciler
+	ReconcilerName = "PubSub"
+)
+
+// Reconciler implements controller.Reconciler for Channel resources.
+type PubSubBase struct {
+	*reconciler.Base
+
+	TopicOpsImage        string
+	SubscriptionOpsImage string
+}
+
+func (c *PubSubBase) ensureTopic(ctx context.Context, channel *v1alpha1.Channel, topicID string) (bool, error) {
+	return false, errors.New("not implemented")
+}
+
+type OpsState string
+
+const (
```

Comment:
I'm slightly concerned about usin a state machine (vs goal-based reconciliation).

On the other hand, these don't actually look to be used as a state machine, just a regular enum. Maybe `OpStatus` and remove `State` from the names?

Reply:
Sure. I will do that in a later PR.